### PR TITLE
Check mappings ONNX -> Caffe2 bear the same argument names

### PR DESCRIPTION
**Pull request by huitseeker**
When adding an extra arg to an input ONNX op, if it's not supported in Caffe2, the exporter would just silently pass it to NetDef and ignore it in the implementation. It's pretty error-prone. Caffe2 also has an OpSchema description and we can enforce that all arguments explicitly appear in schema or listed explicitly in Caffe2.
Adds test for C2 argument checking.

See also https://github.com/caffe2/caffe2/pull/2478

review by @bddppq

### CODE REVIEW

File: caffe2/python/onnx/tests/c2_ref_test.py [hunk 1]
```diff
@@ -27,13 +27,29 @@
...
+        Z = np.zeros((3, 2)).astype(np.float32)
+        
+        b2 = C.Caffe2Backend()
+
+        node_def = make_node("Add", inputs = ["X", "Y"], outputs = ["Z"], broadcast = 0)
+        output = b2.convert_node(node_def.SerializeToString(), 6)
+
+        bad_node_def = make_node("Add", inputs = ["X", "Y"], outputs = ["Z"], foo = 42, bar = 56)
+        with self.assertRaises(RuntimeError):
+            output = b2.convert_node(bad_node_def.SerializeToString(), 6)
```
Comment (bddppq) [inline caffe2/python/onnx/tests/c2_ref_test.py]: could you do a check of the error message here (to make sure we caught the right exception)?

File: caffe2/onnx/backend.h [hunk 2]
```diff
@@ -125,6 +125,8 @@ class Caffe2Backend {
       bool include_initializers,
       const std::vector<Caffe2Ops>& extras);
 
+  void CheckOpSchemaArguments(const caffe2::OpSchema* schema, const Caffe2Ops* op);
```
Comment (bddppq) [inline caffe2/onnx/backend.h]: nit: could you change the arguments to be const ref here?

File: caffe2/onnx/backend.cc [hunk 3]
```diff
@@ -890,17 +890,47 @@ Caffe2Ops Caffe2Backend::ConvertNode(
...
   if (get_special_operators().count(onnx_node->node.op_type())) {
-    return (this->*get_special_operators().at(onnx_node->node.op_type()))(
+    res = (this->*get_special_operators().at(onnx_node->node.op_type()))(
         onnx_node, opset_version);
   } else {
-    return CommonOnnxNodeToCaffe2Ops(onnx_node, opset_version);
+    res = CommonOnnxNodeToCaffe2Ops(onnx_node, opset_version);
   }
+  const auto op_type = onnx_node -> node.op_type();
+  const auto* schema = OpSchemaRegistry::Schema(op_type);
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: schema can be nullptr here if there is no such operator.

Comment (bddppq) [inline caffe2/onnx/backend.cc]: here should lookup the schema of the c2 op, not the onnx node (their name can be different).

File: caffe2/onnx/backend.cc [hunk 4]
```diff
@@ -890,17 +890,47 @@ Caffe2Ops Caffe2Backend::ConvertNode(
   return OnnxNodeToCaffe2Ops(init_model, pred_model, &onnx_node, opset_version);
 }
 
+void Caffe2Backend::CheckOpSchemaArguments(
+    const caffe2::OpSchema* schema,
+    const Caffe2Ops* op) {
+  const auto schema_args = schema -> args();
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: 'const&', no need to make a copy.

File: caffe2/onnx/backend.cc [hunk 5]
```diff
@@ -890,17 +890,47 @@ Caffe2Ops Caffe2Backend::ConvertNode(
...
+  std::vector<std::string> argnames;
+  std::transform(
+      schema_args.begin(),
+      schema_args.end(),
+      std::back_inserter(argnames),
+      [](caffe2::OpSchema::Argument elem) { return elem.name(); });
+
+  for (const auto& result_op : op -> ops) {
+    for (const auto& arg : result_op.arg()) {
+      if (std::find(argnames.begin(), argnames.end(), arg.name()) ==
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: nit: you can do `argnames.count` here.

File: caffe2/onnx/backend.cc [hunk 6]
```diff
@@ -890,17 +890,47 @@ Caffe2Ops Caffe2Backend::ConvertNode(
...
+          argnames.end()) {
+        CAFFE_THROW(
+            "Don't know how to map unexpected argument ",
+            arg.name(),
+            " (from operation ",
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: nit: operation -> operator

**Review by bddppq — COMMENTED** [review]
Thanks for adding the check

**Comment by huitseeker** [comment]
Thanks for the review, @bddppq

File: caffe2/onnx/backend.cc [hunk 7]
```diff
@@ -890,17 +890,50 @@ Caffe2Ops Caffe2Backend::ConvertNode(
   return OnnxNodeToCaffe2Ops(init_model, pred_model, &onnx_node, opset_version);
 }
 
+void Caffe2Backend::CheckOpSchemaArguments(
+    const caffe2::OpSchema& schema,
+    const caffe2::OperatorDef& op) {
+  const auto schema_args = schema.args();
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: const auto&

File: caffe2/onnx/backend.cc [hunk 8]
```diff
@@ -890,17 +890,50 @@ Caffe2Ops Caffe2Backend::ConvertNode(
...
+    res = (this->*get_special_operators().at(onnx_node->node.op_type()))(
         onnx_node, opset_version);
   } else {
-    return CommonOnnxNodeToCaffe2Ops(onnx_node, opset_version);
+    res = CommonOnnxNodeToCaffe2Ops(onnx_node, opset_version);
+  }
+
+  for (const auto& result_op: res.ops){
+    const auto* schema = OpSchemaRegistry::Schema(result_op.type());
+    if (true) {
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: `if (schema)`?

**Review by bddppq — APPROVED** [review]
LG, just some small places to fix before merging.

File: caffe2/onnx/backend.cc [hunk 9]
```diff
@@ -890,17 +890,50 @@ Caffe2Ops Caffe2Backend::ConvertNode(
...
+      std::back_inserter(argnames),
+      [](caffe2::OpSchema::Argument elem) { return elem.name(); });
+
+  for (const auto& arg : op.arg()) {
+    if (std::count(argnames.begin(), argnames.end(), arg.name()) == 0) {
+      CAFFE_THROW(
+          "Don't know how to map unexpected argument ",
+          arg.name(),
+          " (from operator ",
+          op.name(), ")");
```
Comment (bddppq) [inline caffe2/onnx/backend.cc]: Here better to print `op.type()`

**Comment by bddppq** [comment]
The test failure is caused by caffe2 GivenTensor.*Fill operators have not correctly declared their arguments. I will send a separate PR for fixing that, this PR needs to wait for it.

**Comment by huitseeker** [comment]
@bddppq I think I can work around this failure with this trick: https://github.com/pytorch/pytorch/pull/6317/files#diff-b8145021cec305709a7612a7f32603d8R898

**Comment by bddppq** [comment]
@huitseeker the fix is in master now, could you revert the hack so we can do the real testing?

**Pull request merged**